### DISCUSSION TITLE
How to return an error from an extension?

### DISCUSSION THREAD

**Discussion by otryshko** [mirumee/ariadne#749] [Q&A]
How do I bubble an exception out from extension? If I just raise one, request fails with a 500; I'd prefer the exception to be present in the errors node of the standard graphql response payload.

**Comment by rafalp** [comment]
Can you share what is your usecase? Generally you shouldn't use extension system for any heavy logic that may be raising exceptions.

**Reply by otryshko** [reply]
@rafalp Authentication is the simplest case - if the request auth header is invalid, send back corresponding error.

**Reply by rafalp** [reply]
You are not supposed to use extension system for authentication. I've explained this more [here](https://github.com/mirumee/ariadne/issues/710).

**Reply by otryshko** [reply]
@rafalp I assume you suggest pushing auth to the hosting runtime/middleware? Then we'll have to replicate graphql data formatter there (that `data.errors` json blob). It would be nice to be able to reuse an existing exception -> json code path

**Reply by rafalp** [reply]
`data.errors` is not created by error formatter, only `result.errors` which is level above.

Replicating formatter here should be as simple as returning JSON response with `{"data": None, "errors": [format_error(any_error)]}`, but relying on `result.errors` is popularly considered less productive approach than using custom error messaging in schema.
